src/utils/addons/prepare.js

```diff
@@ -0,0 +1,110 @@
+const chalk = require('chalk')
+
+const ADDON_VALIDATION = {
+  EXISTS: 'EXISTS',
+  NOT_EXISTS: 'NOT_EXISTS',
+}
+
+const validateExists = ({ addon, addonName, siteData, log, exit }) => {
+  if (!addon || !addon.id) {
+    log(`Add-on ${addonName} doesn't exist for ${siteData.name}`)
+    log(`> Run \`netlify addons:create ${addonName}\` to create an instance for this site`)
+    exit(1)
+  }
+}
+
+const validateNotExists = ({ addon, addonName, siteData, log, exit }) => {
+  if (addon && addon.id) {
+    log(`The "${addonName} add-on" already exists for ${siteData.name}`)
+    log()
+    const cmd = chalk.cyan(`\`netlify addons:config ${addonName}\``)
+    log(`- To update this add-on run: ${cmd}`)
+    const deleteCmd = chalk.cyan(`\`netlify addons:delete ${addonName}\``)
+    log(`- To remove this add-on run: ${deleteCmd}`)
+    log()
+    exit(1)
+  }
+}
+
+const getCurrentAddon = ({ addons, addonName }) => {
+  const addon = addons.find(addon => addon.service_path === `/.netlify/${addonName}`)
+  return addon
+}
+
+const validateCurrentAddon = ({ addon, validation, addonName, siteData, log, warn, exit }) => {
+  switch (validation) {
+    case ADDON_VALIDATION.EXISTS: {
+      validateExists({ addon, addonName, siteData, log, exit })
+      break
+    }
+    case ADDON_VALIDATION.NOT_EXISTS: {
+      validateNotExists({ addon, addonName, siteData, log, exit })
+      break
+    }
+    default: {
+      warn(`Unknown addons validation: ${validation}`)
+      break
+    }
+  }
+}
+
+const getAddonManifest = async ({ api, addonName, error }) => {
+  let manifest
+  try {
+    manifest = await api.showServiceManifest({ addonName })
+  } catch (e) {
+    if (typeof e.message === 'string' && e.message.includes('Not Found')) {
+      error(`No add-on "${addonName}" found. Please double check your add-on name and try again`)
+    } else {
+      error(e.message)
+    }
+  }
+  return manifest
+}
+
+const getSiteData = async ({ api, siteId, error }) => {
+  let siteData
+  try {
+    siteData = await api.getSite({ siteId })
+  } catch (e) {
+    error(`Failed getting list of site data: ${e.message}`)
+  }
+  return siteData
+}
+
+const getAddons = async ({ api, siteId, error }) => {
+  let addons
+  try {
+    addons = await api.listServiceInstancesForSite({ siteId })
+  } catch (e) {
+    error(`Failed getting list of addons: ${e.message}`)
+  }
+  return addons
+}
+
+const prepareAddonCommand = async ({ context, addonName, validation }) => {
+  const { authenticate, netlify, log, warn, error, exit } = context
+  const { api, site } = netlify
+  const siteId = site.id
+  if (!siteId) {
+    error('No site id found, please run inside a site folder or `netlify link`')
+  }
+
+  await authenticate.call(context)
+
+  const [manifest, siteData, addons] = await Promise.all([
+    addonName ? getAddonManifest({ api, addonName, error }) : Promise.resolve(),
```

Comment:
Some of the functions calling this function only need one or two of the `manifest`, `siteData` and `addons`. This means they would make unnecessary API calls. Should we optimize to remove those?

We could for example pass an array option to specify which data should be retrieved.

Another way could also be to split this function into three different functions and let the caller do the `await Promise.all([...])`. `context.authenticate()` could be called from within `getAddonManifest()`, etc. so the caller does not need to call it, but it would need to be memoized then.

Reply:
All callers need the call to get `addons`. `list` needs it to display `addons`, but all others require it for validation.
The `manifest` call has a side effect of verifying the user didn't make a typo.
For example `addons:delete non-existing`, would previously print:
```
Add-on non-existing doesn't exist for <site-name>
> Run `netlify addons:create non-existing` to create an instance for this site`
```
Since it was verifying the addon exists in the list of installed addons.

Now it will print:
```
No add-on "non-existing" found. Please double check your add-on name and try again
```

Reply:
Oh I see, I did not think of the side effect of `manifest` validation. 
Then it makes all sense :+1: